### PR TITLE
add delay_list to tag delay init_by_lua or not

### PR DESCRIPTION
**Pull request by rainingmaster**
Some time we need to delayed execution of init_by_lua. For example, a third party module using `ngx_shm_zone_t`, and also want to be able to use this module in init_by_lua.
So I think we should provide a init_by_lua to the other third party module for callback.

### CODE REVIEW

**Comment by agentzh** [comment]
@rainingmaster The 3rd-party module should use ngx_lua's public C API to register their shm zones. See

https://github.com/openresty/lua-nginx-module/blob/master/src/api/ngx_http_lua_api.h

And an example module here:

https://github.com/openresty/lua-nginx-module/tree/master/t/data/fake-shm-module

**Comment by rainingmaster** [comment]
@agentzh We encountered a situation, we hope that using the shm in `http lua` and `stream lua`  at the same time, so the `init_by_lua` phase of the above two modules need to be ran  after `*_init_zone`. Therefore, it is desirable to have a callback function to control the execution time of `init_by_lua`

**Comment by agentzh** [comment]
Okay, I see what problem you are trying to solve.

File: src/api/ngx_http_lua_api.h [hunk 1]
```diff
@@ -49,6 +49,11 @@ ngx_shm_zone_t *ngx_http_lua_find_zone(u_char *name_data, size_t name_len);
 ngx_shm_zone_t *ngx_http_lua_shared_memory_add(ngx_conf_t *cf, ngx_str_t *name,
     size_t size, void *tag);
 
+ngx_int_t ngx_http_lua_delay_list_register(ngx_conf_t *cf,
```
Comment (agentzh) [inline src/api/ngx_http_lua_api.h]: We need a better name. "delay list" is very confusing.

File: src/ngx_http_lua_api.c [hunk 2]
```diff
@@ -194,13 +198,12 @@ ngx_http_lua_shared_memory_init(ngx_shm_zone_t *shm_zone, void *data)
 
     lmcf->shm_zones_inited++;
 
-    if (lmcf->shm_zones_inited == lmcf->shm_zones->nelts
-        && lmcf->init_handler)
+    if (lmcf->shm_zones_inited == lmcf->shm_zones->nelts)
```
Comment (agentzh) [inline src/ngx_http_lua_api.c]: Why did you remove the check for an init handler here?

File: src/ngx_http_lua_api.c [hunk 3]
```diff
@@ -213,4 +216,75 @@ ngx_http_lua_shared_memory_init(ngx_shm_zone_t *shm_zone, void *data)
...
+    if (lmcf->delay_list->nelts > lmcf->delay_list_inited) {
+        return NGX_ERROR;
+    }
+
+    if (lmcf->delay_list->nelts < lmcf->delay_list_inited) {
+        return NGX_OK;
+    }
+
+    /* lmcf->delay_list->nelts == lmcf->delay_list_inited */
+     return lmcf->init_handler(lmcf->cycle->log, lmcf, lmcf->lua);
```
Comment (agentzh) [inline src/ngx_http_lua_api.c]: Indentation issue?

File: src/ngx_http_lua_api.c [hunk 4]
```diff
@@ -213,4 +216,75 @@ ngx_http_lua_shared_memory_init(ngx_shm_zone_t *shm_zone, void *data)
...
+ngx_http_lua_delay_init_phase()
+{
+    ngx_http_lua_main_conf_t     *lmcf;
+
+    lmcf = ngx_http_cycle_get_module_main_conf(ngx_cycle, ngx_http_lua_module);
+    if (lmcf == NULL) {
+        return NGX_ERROR;
+    }
+
+    if (! lmcf->init_handler) {
```
Comment (agentzh) [inline src/ngx_http_lua_api.c]: Style: we never put a space after the prefix operator `!`. Please fix other similar places as well.

File: src/ngx_http_lua_api.c (hunk 2 above)
Comment (rainingmaster) [inline src/ngx_http_lua_api.c]: @agentzh because I think use like a third module here: they can not get `lmcf->init_handler`,  and we can check init handler in `ngx_http_lua_delay_init_phase` api

File: src/ngx_http_lua_api.c (hunk 4 above)
Comment (rainingmaster) [inline src/ngx_http_lua_api.c]: @agentzh OK, I will fix it. Is there the same problem in https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_args.c#L253 and https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_capturefilter.c#L60, do I need to fix them? Or we just put a space before parameter in the second paragraph of judgment

**Comment by rainingmaster** [comment]
@agentzh I am sorry so much style issue in my code... I've try to fix them. But I am not good at name the code, can you help me to name such as `delay_list`, `delay_list_init` and `ngx_http_lua_delay_list_register`, `ngx_http_lua_delay_init_phase`.
And [here](https://github.com/openresty/lua-nginx-module/pull/1050/files#diff-3ed27dd07994dffb9efa8c835e198a15R262), it is safe enough to use `ngx_cycle` for other modules?
Thanks

File: src/ngx_http_lua_api.c (hunk 4 above)
Comment (agentzh) [inline src/ngx_http_lua_api.c]: @rainingmaster Yeah, those fixes should be in a separate PR. Each PR needs to be self-contained.

File: src/ngx_http_lua_api.c [hunk 5]
```diff
@@ -213,4 +216,74 @@ ngx_http_lua_shared_memory_init(ngx_shm_zone_t *shm_zone, void *data)
...
+        return NGX_ERROR;
+    }
+
+    if (!lmcf->init_handler) {
+        return NGX_OK;
+    }
+
+    lmcf->delay_list_inited++;
+
+    if (lmcf->delay_list->nelts > lmcf->delay_list_inited) {
```
Comment (agentzh) [inline src/ngx_http_lua_api.c]: What does this mean?

File: src/ngx_http_lua_api.c [hunk 6]
```diff
@@ -213,4 +216,74 @@ ngx_http_lua_shared_memory_init(ngx_shm_zone_t *shm_zone, void *data)
...
+        return NGX_ERROR;
+    }
+
+    if (lmcf->delay_list->nelts < lmcf->delay_list_inited) {
```
Comment (agentzh) [inline src/ngx_http_lua_api.c]: What does this mean?

File: src/ngx_http_lua_api.c [hunk 7]
```diff
@@ -213,4 +216,74 @@ ngx_http_lua_shared_memory_init(ngx_shm_zone_t *shm_zone, void *data)
     return NGX_OK;
 }
 
+
+ngx_int_t
+ngx_http_lua_delay_list_register(ngx_conf_t *cf, void *tag)
```
Comment (agentzh) [inline src/ngx_http_lua_api.c]: What exactly is being registered here? Modules requiring delayed init handler invocation?

File: src/ngx_http_lua_common.h [hunk 8]
```diff
@@ -176,6 +178,9 @@ struct ngx_http_lua_main_conf_s {
 
     ngx_array_t         *preload_hooks; /* of ngx_http_lua_preload_hook_t */
 
+    ngx_array_t         *delay_list; /* of void* */
+    unsigned             delay_list_inited;
```
Comment (agentzh) [inline src/ngx_http_lua_common.h]: I don't know what these fields' purpose is.

File: src/ngx_http_lua_api.c [hunk 9]
```diff
@@ -194,13 +198,12 @@ ngx_http_lua_shared_memory_init(ngx_shm_zone_t *shm_zone, void *data)
...
 
-    if (lmcf->shm_zones_inited == lmcf->shm_zones->nelts
-        && lmcf->init_handler)
+    if (lmcf->shm_zones_inited == lmcf->shm_zones->nelts)
     {
         saved_cycle = ngx_cycle;
         ngx_cycle = ctx->cycle;
 
-        rc = lmcf->init_handler(ctx->log, lmcf, lmcf->lua);
+        rc = ngx_http_lua_delay_init_phase();
```
Comment (agentzh) [inline src/ngx_http_lua_api.c]: What happens when there is no `lua_shared_dict` defined in `nginx.conf`? Is it delayed or not?

Comment (rainingmaster) [inline src/ngx_http_lua_api.c]: @agentzh It will not delayed. Because no one use `ngx_http_lua_delay_list_register` to declare need to delay `init_by_lua`, and `init_by_lua` will still run [here](https://github.com/openresty/lua-nginx-module/pull/1050/files#diff-c91b2ec425e9be9d56fee2df662b67d2R745)

File: src/ngx_http_lua_api.c (hunk 7 above)
Comment (rainingmaster) [inline src/ngx_http_lua_api.c]: @agentzh Yes, I make it like `required_shm` to mark delayed init handler, and we can run it in a same module many times.

File: src/ngx_http_lua_api.c (hunk 5 above)
Comment (rainingmaster) [inline src/ngx_http_lua_api.c]: @agentzh Because `ngx_http_lua_delay_init_phase` will  run in third part module, I think that if there are third party modules called `ngx_http_lua_delay_init_phase` too much times, that is, the number of times the `init` more than the number of `register`, the error should be returned.

File: src/ngx_http_lua_api.c (hunk 6 above)
Comment (rainingmaster) [inline src/ngx_http_lua_api.c]: @agentzh `ngx_http_lua_delay_init_phase` will run in different third part module. We will compare the number of times the init more than the number of the register: if `init` more than `register`, it is an error; if `init` less than `register`, wait other module run `ngx_http_lua_delay_init_phase`; if `init` equal `register`, run `init_by_lua`.

File: src/ngx_http_lua_common.h (hunk 8 above)
Comment (rainingmaster) [inline src/ngx_http_lua_common.h]: @agentzh  `delay_list` is used for register all third part module which need to delay `init_by_lua`, `delay_list_inited` is used for count how many module has run `ngx_http_lua_delay_init_phase`, and used to determine whether `init_by_lua` should be executed

File: src/ngx_http_lua_api.c (hunk 6 above)
Comment (agentzh) [inline src/ngx_http_lua_api.c]: @rainingmaster Then we need an error log message instead of simply returning an `NGX_ERROR`. The current way of handling this makes it very hard to debug this problem.

File: src/ngx_http_lua_common.h (hunk 8 above)
Comment (agentzh) [inline src/ngx_http_lua_common.h]: @rainingmaster I think you should add these as code comments to improve code readability. Also, I think we definitely need better names for these fields.

I don't think we should count the number of registrations. We only need to know whether or not to delay. Your registration API does not really register any real data from the 3rd-party modules. No callbacks from the 3rd-party modules are registered here anyway.

Comment (agentzh) [inline src/ngx_http_lua_common.h]: @rainingmaster Using an array to count registrations is also very strange to me. It just does not make any sense. If you really want to count registrations, we should just maintain a single counter instead of an array without any useful data in its elements. Am I overlooking anything here?

Comment (agentzh) [inline src/ngx_http_lua_common.h]: Also, please use the term "delay_init_handler" in your identifier names. Simply saying "delay" is very confusing.

**Comment by agentzh** [comment]
Let me repeat in case you missed it:

We need a test module under `t/` and some tests to cover these new C APIs.

File: src/ngx_http_lua_common.h (hunk 8 above)
Comment (rainingmaster) [inline src/ngx_http_lua_common.h]: @agentzh using an array here because that I just want to make it like `required_shm`, we can run it again and again in a same third part module, and it is ported from `ngx_shared_memory_add`.
Maybe we just use a counter here and force mandatory third part module just run once in their own init phase?

**Comment by rainingmaster** [comment]
@agentzh OK, I will add the test case as soon as possible.
Because I think of a situation: if a third party module needs to ensure that their postconfiguration run before `init_by_lua` (may need to use some of the variables in `init_by_lua` which defined in the third part module), we may also use this module. But I found that Lua can not use `ngx.var.*` to get this information in `init_by_module`, how can I improve this use case. Or only a third part `ngx_shm` is more appropriate?

**Comment by agentzh** [comment]
@rainingmaster You know that NGINX variables are only available for requests? In those init contexts, there are not nginx requests yet, obviously.

File: src/ngx_http_lua_common.h (hunk 8 above)
Comment (agentzh) [inline src/ngx_http_lua_common.h]: @rainingmaster What do you mean by "third party module just run once in their own init phase"?

**Comment by rainingmaster** [comment]
@agentzh Oh， I know. In addition to using `ngx_shm`, I have other ways  to test it ?

**Comment by agentzh** [comment]
@rainingmaster See the existing test modules under `t/data/`. We only need a minimal nginx C module for testing these APIs. It can be dummy and not really functional. It's all about testing and code coverage.

File: src/ngx_http_lua_common.h (hunk 8 above)
Comment (rainingmaster) [inline src/ngx_http_lua_common.h]: @agentzh I have use (`ngx_http_lua_delay_init_declare`)[https://github.com/openresty/lua-nginx-module/pull/1050/files#diff-3ed27dd07994dffb9efa8c835e198a15R243] to check that every third part module has register or not. 
If we only use a number as counter instead of array `delay_list`, we have to ensure the third part module register only not more than twice.

File: src/ngx_http_lua_api.c [hunk 10]
```diff
@@ -194,13 +198,12 @@ ngx_http_lua_shared_memory_init(ngx_shm_zone_t *shm_zone, void *data)
...
 
-    if (lmcf->shm_zones_inited == lmcf->shm_zones->nelts
-        && lmcf->init_handler)
+    if (lmcf->shm_zones_inited == lmcf->shm_zones->nelts)
     {
         saved_cycle = ngx_cycle;
         ngx_cycle = ctx->cycle;
 
-        rc = lmcf->init_handler(ctx->log, lmcf, lmcf->lua);
+        rc = ngx_http_lua_delay_init_handler();
```
Comment (agentzh) [inline src/ngx_http_lua_api.c]: When there is no `lua_shared_dict` directive configured in `nginx.conf`, there the delayed execution will never run, am I correct? This looks wrong to me.

Comment (rainingmaster) [inline src/ngx_http_lua_api.c]: @agentzh YES. If there is no `lua_shared_dict`, or we can say there is no one declare delay `init_by_lua`(I've moved all delay `init_by_lua to use ngx_http_lua_delay_init_handler to implement, include original `lua_shared_dict`). Then, the `lmcf->delay_init` is `null`, `init_by_lua` will run in ngx_lua's postconfiguration, that is [here](https://github.com/openresty/lua-nginx-module/pull/1050/files#diff-c91b2ec425e9be9d56fee2df662b67d2R746).

**Comment by rainingmaster** [comment]
@agentzh I have try to add a test case to test the `ngx_http_lua_delay_init_handler` api. 
However, the ngx_cycle is controlled by third part module [here](https://github.com/openresty/lua-nginx-module/pull/1050/files#diff-b4a8646df8a3d1d284dc4ab1916cbfbbR214), that may cause some problem to `init_by_lua`, this is not appropriate?

File: src/ngx_http_lua_api.c (hunk 10 above)
Comment (agentzh) [inline src/ngx_http_lua_api.c]: @rainingmaster That does not sound right to me. The 3rd-party modules can do whatever it likes, which may have nothing to do with ngx_lua's own `lua_shared_dict` directive. Or your change here is only meant for `ngx_meta_lua_module` and `lua_shared_dict`? That is, not meant for general 3rd-party modules?

Comment (rainingmaster) [inline src/ngx_http_lua_api.c]: @agentzh The 3rd-part won't relay on the ngx_lua's own `lua_shared_dict`, they can use `ngx_http_lua_delay_init_handler` in their own module like [this](https://github.com/openresty/lua-nginx-module/pull/1050/files#diff-b4a8646df8a3d1d284dc4ab1916cbfbbR217), and `init_by_lua` will be delayed. 
But the disadvantage is that the third party module may need to maintain a `ngx_cycle`.

Comment (agentzh) [inline src/ngx_http_lua_api.c]: @rainingmaster It's the 3rd-party module's responsibility to call ngx_lua's `init_by_lua*` code? Hmm, this looks very hacky and fragile to me.

**Comment by agentzh** [comment]
I'm afraid this PR's approach has inherent problems in its direction.

The model should be like this:

* `ngx_meta_lua_module` implements the shdict (and other things).
* `ngx_meta_lua_module` can work with `ngx_stream_lua_module` alone, or with `ngx_http_lua_module` alone, or both, or neither.
* `ngx_http_lua_module` and `ngx_stream_lua_module` both *register* into `ngx_meta_lua_module`, but not the reversed direction.
* any 3rd-party nginx C modules who want to implement custom shm-based stuff to be shared between `ngx_http_lua_module` and `ngx_stream_lua_module` should register *only* to the `ngx_meta_lua_module`.

I'm afraid this PR fails to achieve the goal in a clean enough way. We should never expose bloody implementation details like "delay init handler" to the public C API for 3rd-party modules, which is too fragile, too hacky, and breaks encapsulation. Also, it's not scalable nor flexible.
